### PR TITLE
Support diff mode in koji_tag and koji_target modules

### PR DESCRIPTION
**Pull request by hongweiliu17**

### CODE REVIEW

**Comment by ktdreyer** [comment]
Great work, thank you!

`tox` fails on py27 and py36, and I've fixed that in https://github.com/ktdreyer/koji-ansible/pull/254 . Would you please rebase onto that?

I've added you as a contributor so you should be able to run the integration tests automatically now.

Regarding tag inheritance: Let's stop printing the "current/new inheritance" in `stdout_lines`. That was giving us too much detail that is better suited for diff mode.

**Comment by hongweiliu17** [comment]
@ktdreyer  Thank you Ken!
I rebased the code and added several unittests for my added code. Do we need to make the code are 100% covered in unittests?
And I can't find integration-tests is run for this PR. Could you take a look?

**Comment by simonbaird** [comment]
Looks good. My only comment is this:

For the intializing of result, e.g. like this:
```
result = {'changed': False, 'stdout_lines': [], 'diff': {}}
```

In the errata tool modules I didn't populate the diff key with an empty dict, so for a task with no change the key would not be present. (I don't know if it's better or worse to have it there as an empty dict, but just noting that it's different. Maybe we should do it like that for the errata modules too.)

**Comment by hongweiliu17** [comment]
Thanks @simonbaird I removed the empty diff from the result.

**Comment by hongweiliu17** [comment]
The test pass now. Thanks.

**Comment by hongweiliu17** [comment]
@ktdreyer  @simonbaird Can you review again and is there any test needed to merge it? Thanks.

**Comment by ktdreyer** [comment]
Thanks for getting to the passing tests. I've not tried this out yet, but I will do it this week. Here are some things I noticed as I read this code:

- [ ] Please fix the `pcakges` -> `packages` spelling

- [ ] Please remove the change that alters `stdout` for check mode. It's not directly related to this PR, and I'd like to preserve the existing behavior for consistency and simplicity. Both check mode and non-check mode should return the same `deleted tag` or `deleted target` messages.
  - Fewer codepaths makes this code easier to unit test.
  - When we always return the same result strings, it's easier for playbook authors to conditionally act on those results. For example, `when result.stdout ...` conditions in playbooks will be the same whether in check mode or real mode.

- [ ] I need to verify that all the new fixture additions really match Koji Hub's API. I didn't get to this today.

- [ ] In the tests, why do we have to call `sort()` on the resulting and expected `stdout_lines`?

I will try this branch out this week and report back here.

**Comment by hongweiliu17** [comment]
> Thanks for getting to the passing tests. I've not tried this out yet, but I will do it this week. Here are some things I noticed as I read this code:
> 
>     * [ ]  Please fix the `pcakges` -> `packages` spelling

Thanks, this is fixed.
> 
>     * [ ]  Please remove the change that alters `stdout` for check mode. It's not directly related to this PR, and I'd like to preserve the existing behavior for consistency and simplicity. Both check mode and non-check mode should return the same `deleted tag` or `deleted target` messages.
>       
>       * Fewer codepaths makes this code easier to unit test.
>       * When we always return the same result strings, it's easier for playbook authors to conditionally act on those results. For example, `when result.stdout ...` conditions in playbooks will be the same whether in check mode or real mode.
> 
Change it back, sorry, I thought the check mode will not really make changes.

>     * [ ]  I need to verify that all the new fixture additions really match Koji Hub's API. I didn't get to this today.
> 
It is very appreciated. I als ocalled koji Hub's api to see the the result of API.
>     * [ ]  In the tests, why do we have to call `sort()` on the resulting and expected `stdout_lines`?
> 
I add sort() to make the tests pass in py27.
I found in py27, the two list with the same elements but different sequence of elements will not be regarded as equivalent lists.
In py36 and py39, the comparison works without sort().
> 
> I will try this branch out this week and report back here.
Thank you!

**Comment by ktdreyer** [comment]
Great, thanks.

Today I verified that the new `FakeKojiSession` methods align with [Koji's API](https://koji.fedoraproject.org/koji/api). The parameter names are different in the fake here, and that's ok for now. It's something we can fix later in a separate PR.

I looked at building integration tests for this. I found out that modules do not return diff data at all unless we invoke `ansible-playbook` with the `--diff` option. Today the integration tests do not do that, so it's difficult to assert the result values from tasks unless we add that `--diff` to the integration tests. Maybe it's simplest to just add it to all the tests.

Looks like we get some spurious diffs with the `extra` field when I change something else on a tag (for example, when I change the package list). The diffs look like this:

```diff
-extra: {}
+extra: null
```

Maybe we should normalize that somehow so it is clear that `extra` is not really changing.

I'm continuing to test this further.

**Comment by hongweiliu17** [comment]
> Great, thanks.
> 
> Today I verified that the new `FakeKojiSession` methods align with [Koji's API](https://koji.fedoraproject.org/koji/api). The parameter names are different in the fake here, and that's ok for now. It's something we can fix later in a separate PR.
> 

Let me verify them.

> I looked at building integration tests for this. I found out that modules do not return diff data at all unless we invoke `ansible-playbook` with the `--diff` option. Today the integration tests do not do that, so it's difficult to assert the result values from tasks unless we add that `--diff` to the integration tests. Maybe it's simplest to just add it to all the tests.
> 

Yes, let me add --diff to run.sh, let's see

> Looks like we get some spurious diffs with the `extra` field when I change something else on a tag (for example, when I change the package list). The diffs look like this:
> 
> ```diff
> -extra: {}
> +extra: null
> ```
> 
> Maybe we should normalize that somehow so it is clear that `extra` is not really changing.

Yes, I should generate diff data for extra only when 'extra' exists in the params of koji_tag.

> 
> I'm continuing to test this further.

Thank you very much for your time!

**Comment by hongweiliu17** [comment]
I updated the method in FakeKojiSession added by me. But looks we still have some other method to be updated to align with Koji's API

There are a failed test in integration test now. I will fix it tomorrow.

**Comment by ktdreyer** [comment]
I've re-run the tests on `master` with the latest Koji `master` to see if there's a regression in Koji: https://github.com/ktdreyer/koji-ansible/actions/runs/1550694859

That test failed with [a131f8389188809a4135f9d49ffc9414cb2e54cd](https://pagure.io/koji/c/a131f8389188809a4135f9d49ffc9414cb2e54cd). The last one that passed succeeded with [11ffdf6480216be4fe9f6af24ebbc318ac98b751](https://pagure.io/koji/c/11ffdf6480216be4fe9f6af24ebbc318ac98b751) . I suspect there is a regression in Koji here. Next step would be to check out that older Koji sha in [tests/integration/setup.sh](https://github.com/ktdreyer/koji-ansible/blob/master/tests/integration/setup.sh) that passed and re-run again.

I will look more into this problem tomorrow

**Comment by ktdreyer** [comment]
I found [the tests passed](https://github.com/ktdreyer/koji-ansible/actions/runs/1609098795) with [that older Koji sha](https://github.com/ktdreyer/koji-ansible/commit/8bda17627125ebab6c81f07a5712ccdc143112f6), and I reported the regression at https://pagure.io/koji/issue/3193 .

**Comment by hongweiliu17** [comment]
Thank you very much for your debug! I just thought about koji-ansible but didn't think of koji. Learnt it.

**Comment by hongweiliu17** [comment]
@ktdreyer  I checkout the older commit 11ffdf6 in setup.sh and the integration pass now, could you take a look again?

**Review by ktdreyer — COMMENTED** [review]
Looks better, thank you! I see the diff output in the integration tests now!

When I come back to this in January, I will try adding some integration tests to assert these expected `diff` values, like we do now for `stdout_lines`. (errata-tool-ansible's test suite has many `stdout_lines` assertions, but koji-ansible has almost no `stdout_lines` assertions yet.)

File: library/koji_tag.py [hunk 1]
```diff
@@ -493,6 +552,7 @@ def ensure_blocked_packages(session, tag_id, check_mode, packages):
     :returns: a list of human-readable changes
```
[the remark this answers is not in the document]
Comment (hongweiliu17) [reply library/koji_tag.py]: Thanks, It is updated.
I also add a few assertion in integration tests of koji_tag and koji_target.

**Comment by ktdreyer** [comment]
This looks great. Thank you for adding so many integration test improvements!

The changes are squashed together now, and the second commit has many more changes in it.

I've fixed the test suite to just use my upstream Koji fix for now: https://github.com/ktdreyer/koji-ansible/pull/257 Could you please rebase onto that?

**Comment by hongweiliu17** [comment]
@ktdreyer The MR is rebased. And two commits are squashed together as well then it might be cleaner in git log. Thank you!

**Comment by ktdreyer** [comment]
Thank you! I had to address a separate issue today. I am going to run this on the rhel8 playbooks one final time and then merge.

**Review by ktdreyer — APPROVED** [review]
I've tested this with the rhel8 role on a local koji instance and it works fine.

**Comment by hongweiliu17** [comment]
Thank very much for your review and testing Ken and Simon.

**Pull request merged**